File: kepad/src/commonMain/kotlin/wedjets/HeksagonTutcboks.kt
package wedjets

import androidx.compose.foundation.layout.Box
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Outline
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.unit.Density
import androidx.compose.ui.unit.LayoutDirection
import kotlin.math.cos
import kotlin.math.sin

/**
 * A custom shape that clips its content into a hexagon.
 * This is the Compose equivalent of the custom RenderObject used for hit-testing in the original Flutter code.
 * By clipping the shape, both the visual representation and the touch input area are constrained to the hexagon.
 *
 * @param rotationAngle The angle in radians to rotate the hexagon.
 */
class HexagonShape(private val rotationAngle: Float) : Shape {
    override fun createOutline(
        size: Size,
        layoutDirection: LayoutDirection,
        density: Density
    ): Outline {
        return Outline.Generic(
            path = createHexagonPath(size, rotationAngle)
        )
    }

    private fun createHexagonPath(size: Size, rotation: Float): Path {
        val path = Path()
        val centerX = size.width / 2f
        val centerY = size.height / 2f

        // The original RenderObject subtracted a small fixed value to create a "safe zone".
        // We'll use a small percentage of the height for better scaling.
        val safeZone = size.height * 0.02f
        val radius = (size.height / 2f) - safeZone

        for (i in 0..5) {
            // Angle for pointy-top hexagons (-30 degrees offset), converted to radians.
            val angleRad = (i * 60f - 30f) * (Math.PI.toFloat() / 180f) + rotation

            val x = centerX + radius * cos(angleRad)
            val y = centerY + radius * sin(angleRad)

            if (i == 0) {
                path.moveTo(x, y)
            } else {
                path.lineTo(x, y)
            }
        }
        path.close()
        return path
    }

    // Two shapes with the same rotation are considered equal.
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is HexagonShape) return false
        return rotationAngle == other.rotationAngle
    }

    override fun hashCode(): Int {
        return rotationAngle.hashCode()
    }
}

/**
 * A composable that provides a hexagonal touch area for its child.
 * It uses a custom Shape to clip the bounds for both drawing and touch input.
 *
 * @param modifier The modifier to be applied to the layout.
 * @param rotationAngle The angle in radians to rotate the hexagonal clip shape.
 * @param content The content to be displayed inside the hexagonal area.
 */
@Composable
fun HeksagonTutcboks(
    modifier: Modifier = Modifier,
    rotationAngle: Float,
    content: @Composable () -> Unit
) {
    Box(
        modifier = modifier.clip(HexagonShape(rotationAngle))
    ) {
        content()
    }
}

File: lib/widgets/heksagon_tutcboks.dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'dart:math' as math;

class HeksagonTutcboks extends SingleChildRenderObjectWidget {
  final double rotationAngle;

  const HeksagonTutcboks({
    super.key,
    this.rotationAngle = 0.0,
    required super.child,
  });

  @override
  RenderObject createRenderObject(BuildContext context) {
    return RenderHeksagonTutcboks(rotationAngle: rotationAngle);
  }

  @override
  void updateRenderObject(
      BuildContext context, RenderHeksagonTutcboks renderObject) {
    renderObject.rotationAngle = rotationAngle;
  }
}

class RenderHeksagonTutcboks extends RenderProxyBox {
  double rotationAngle;

  RenderHeksagonTutcboks({
    this.rotationAngle = 0.0,
  });

  // Cache path for performance
  Path? _cachedHexPath;
  Size? _cachedSize;
  double? _cachedRotation;

  Path _createHexagonPath(Size size) {
    if (_cachedHexPath != null &&
        _cachedSize == size &&
        _cachedRotation == rotationAngle) {
      return _cachedHexPath!;
    }

    final path = Path();
    final centerX = size.width / 2;
    final centerY = size.height / 2;

    // The painter uses size.height / 2 for the radius. We subtract a small
    // amount to create a "safe zone" and prevent the hit area from
    // bleeding into neighboring hexagons.
    final radius = (size.height / 2) - 2.0;

    for (int i = 0; i < 6; i++) {
      double angle = (i * 60 - 30) * (math.pi / 180); // Pointy-top
      angle += rotationAngle;

      double x = centerX + radius * math.cos(angle);
      double y = centerY + radius * math.sin(angle);

      if (i == 0) {
        path.moveTo(x, y);
      } else {
        path.lineTo(x, y);
      }
    }
    path.close();

    _cachedHexPath = path;
    _cachedSize = size;
    _cachedRotation = rotationAngle;
    return path;
  }

  @override
  bool hitTest(BoxHitTestResult result, {required Offset position}) {
    // Custom check: ensure the point is inside the hexagon path. If not,
    // we return false immediately, and neither this widget nor its children
    // (including the GestureDetector) will receive the event.
    if (!_createHexagonPath(size).contains(position)) {
      return false;
    }
    // If the point is inside our shape, proceed with the default hit-testing
    // logic for our children.
    return super.hitTest(result, position: position);
  }
}
